src/ui/backend/canvas/animate.js

```diff
@@ -427,53 +483,58 @@ var Animator = exports.Animator = Class(Emitter, function () {
 
 		this._elapsed += dt;
 		this.next();
-
-		if (!this._isScheduled) {
-			this._group.onAnimationFinish(this);
-		}
 	};
 
 	this.next = function () {
-		var p,
-			target;
-
-		if (!this._queue[0]) { return; }
-
-		while ((p = this._queue[0])) {
-			var frameFinished = this._elapsed >= p.duration,
-				t = frameFinished ? 1 : this._elapsed / p.duration,
-				tt = p.transition(t);
+		var p = this._queue[0];
+		while (p) {
+			var duration = p.duration;
+			var frameFinished = this._elapsed >= duration;
+			var t = frameFinished ? 1 : this._elapsed / duration;
+			var tt = p.transition(t);
 
 			if (frameFinished) {
-				this._elapsed -= p.duration;
+				this._elapsed -= duration;
 			}
 
-			try {
-				p.exec(tt, t, this._debug);
-				if (p.onTick) { p.onTick.call(p.subject, tt, t); }
-			} finally {
-				// if we haven't modified the queue in a callback, remove the frame if it is finished
-				if (frameFinished && p == this._queue[0]) {
-					this._queue.shift();
-				}
-
-				// if we got paused during a callback or the
-				// frame is not finished yet, don't continue
-				if (!frameFinished || this._isPaused) { return; }
+			p.exec(tt, t, this._debug);
+			p.onTick && p.onTick.call(p.subject, tt, t);
+
+			// remove frame if finished and queue wasn't modified by a callback
+			if (frameFinished && p === this._queue[0]) {
+				this._queue.shift();
 			}
+
+			// if paused during a callback or frame not finished, don't continue
+			if (!frameFinished || this._isPaused) { return; }
+
+			p = this._queue[0];
 		}
 
 		// nothing left in the queue!
 		this._unschedule();
+		this._removeFromGroup();
+	};
+
+	this._addToGroup = function () {
+		if (this.groupID) {
+			groups[this.groupID].add(this);
+		}
+	};
+
+	this._removeFromGroup = function () {
+		if (this.groupID) {
```

Comment:
I've been seeing errors here with ungrouped animations (`Cannot read property 'remove' of undefined`) - since `groupId` defaults to `DEFAULT_GROUP_ID` this appears to be a redundant check, maybe `if (groups[this.groupID]) {` would be better?